source/Mlos.Python/mlos/Spaces/Hypergrid.py

```diff
@@ -50,13 +59,73 @@ def get_dimensions_for_point(self, point, return_join_dimensions=True):
     def random(self, point=None):
         raise NotImplementedError("All subclasses must implement this.")
 
+    @trace()
     def random_dataframe(self, num_samples):
         config_dicts = [
             {dim_name: value for dim_name, value in self.random()}
             for _ in range(num_samples)
         ]
         return pd.DataFrame(config_dicts)
 
+    @trace()
+    def filter_out_invalid_rows(self, original_dataframe: pd.DataFrame, exclude_extra_columns=True) -> pd.DataFrame:
+        """ Returns a dataframe containing only valid rows from the dataframe.
+
+        Valid rows are rows with no NaNs and with values for all dimensions in the required ranges.
+
+        If there are additional columns, they will be preserved.
+
+        :param original_dataframe:
+        :param exclude_extra_columns:
+        :return:
+        """
+        assert set(original_dataframe.columns.values).issuperset(set(self.dimension_names))
+
+        valid_rows_index = None
+        dataframe = original_dataframe[self.dimension_names]
+
+        if not self.is_hierarchical():
+            # Let's exclude any extra columns
+            #
+            valid_rows_index = dataframe.index[dataframe.notnull().all(axis=1)]
+
+            # Now for each column let's filter out the rows whose values are outside the allowed ranges.

```

Comment:
Is there a reason not to throw an error if the data is not in the right space?

Reply:
Random forest passes all the data to the decision trees and each tree has to filter out data that's within its space. We could do this filtering in the random forest class too, but the mechanics would be the same.

Reply:
Hm I'm wondering if we give up on some input checking here. One of the benefits of having such an explicit system for dimensions is that we can check if everything lines up properly and I feel we're not making use of this here. I don't have a super strong opinion though. I guess it depends on where we call this function.